performance_test/src/communication_abstractions/ros2_communicator.hpp

```diff
@@ -0,0 +1,178 @@
+      ros_qos.best_effort();
+    } else if (m_qos.reliability == QOSAbstraction::Reliability::RELIABLE) {
+      ros_qos.reliable();
+    } else {
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+    if (m_qos.durability == QOSAbstraction::Durability::VOLATILE) {
+      ros_qos.durability_volatile();
+    } else if (m_qos.durability == QOSAbstraction::Durability::TRANSIENT_LOCAL) {
+      ros_qos.transient_local();
+    } else {
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+
+    if (m_qos.history_kind == QOSAbstraction::HistoryKind::KEEP_ALL) {
+      ros_qos.keep_all();
+    } else if (m_qos.history_kind == QOSAbstraction::HistoryKind::KEEP_LAST) {
+      ros_qos.keep_last(m_qos.history_depth);
+    } else {
+      throw std::runtime_error("Unsupported QOS!");
+    }
+
+    return ros_qos;
+  }
+
+private:
+  const QOSAbstraction m_qos;
+};
+
+/// Communication plugin interface for ROS 2 for the subscription side.
+template<class Topic>
+class ROS2Communicator : public Communicator
+{
+public:
+  /// The data type to publish and subscribe to.
+  using DataType = typename Topic::RosType;
+
+  /// Constructor which takes a reference \param lock to the lock to use.
+  explicit ROS2Communicator(SpinLock & lock)
+  : Communicator(lock),
+    m_node(ResourceManager::get().ros2_node())
+  {
+    m_ROS2QOSAdapter = std::make_shared<rclcpp::QoS>(ROS2QOSAdapter(m_ec.qos()).get());
+  }
+
+  /**
+   * \brief Publishes the provided data.
+   *
+   *  The first time this function is called it also creates the publisher.
+   *  Further it updates all internal counters while running.
+   *
+   * \param data The data to publish.
+   * \param time The time to fill into the data field.
+   */
+  void publish(DataType & data, const std::chrono::nanoseconds time)
+  {
+    if (!m_publisher) {
+      auto ros2QOSAdapter = m_ROS2QOSAdapter;
+      // Workaround for bug where RMW/FastRPS keeps history of volatile publisher.
+      if (ros2QOSAdapter->get_rmw_qos_profile().durability ==
+        rmw_qos_durability_policy_t::RMW_QOS_POLICY_DURABILITY_VOLATILE)
+      {
+        ros2QOSAdapter->keep_last(std::size_t(10));  // Setting depth to 10 as a safety net.
+      }
+      // End of workaround.
+
+      m_publisher = m_node->create_publisher<DataType>(
+        Topic::topic_name() + m_ec.pub_topic_postfix(), *ros2QOSAdapter);
+    }
+    lock();
+    data.time = time.count();
+    data.id = next_sample_id();
+    increment_sent();  // We increment before publishing so we don't have to lock twice.
+    unlock();
+    m_publisher->publish(data);
+  }
+
+  /// Reads received data from ROS 2 using callbacks
+  virtual void update_subscription() = 0;
+
+  /// Returns the accumulated data size in bytes.
+  std::size_t data_received()
+  {
+    return num_received_samples() * sizeof(DataType);
+  }
+
+protected:
+  std::shared_ptr<rclcpp::Node> m_node;
+  std::shared_ptr<rclcpp::QoS> m_ROS2QOSAdapter;
```

Comment:
Why is this a shared ptr?

Reply:
No need . can be changed to simple object of type `rclcpp::QoS`